File: blind75/linkedList/206-reverseLinkedList.js
// 206. Reverse Linked List
// Given the head of a singly linked list, reverse the list, and return the reversed list.

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    // 递归的话
    // 考虑几种情况，第一 head === null 第二 head.next === null
    // 第一种情况直接返回null
    if (head === null) return head; 

    let newHead = head;

    if (head.next !== null) {
        // newHead 代表着之后的返回值
        // reverse head之后的listNode
        newHead = reverseList(head.next);
        // head.next 是 当前的下个node T
        // T.next = head 代表着倒叙
        head.next.next = head;
        // head.next 如果不变会成环
        head.next = null;
    }

    // 第二种情况 直接返回head
    return newHead;
}


// var reverseList = function(head) {
//     // iterative
//     let prev = null;
//     let curr = head;
//     while (curr != null) {
//         let tmp = curr.next;
//         curr.next = prev;
//         prev = curr;
//         curr = tmp;
//     }
//     return prev;
// };


let head = [1,2,3,4,5]
let head1 = [1,2];
let head2 = [];
console.log(reverseList(head));
console.log(reverseList(head1));
console.log(reverseList(head2));

File: specified-learning专项训练/linkedList/206-reverseLinkedList.py
# 206. Reverse Linked List
# Given the head of a singly linked list, reverse the list, and return the reversed list.

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

from typing import Optional

class Solution:
    # 递归 递归的办法
    # 有两种特殊情况，
    # 一是当它本身为None的时候，就直接返回None，无法反转
    # 再者当它的next为None的时候，代表着这是最后一个Node
    # 需要返回这个Node，这个Node将为新的head
    # 普通的情况，我们遍历到一个node 
    # node -> aNode <- bNode <-...<-newHead
    # 我们需要 node.next 为 aNode 
    # node.next.next = node 意味着让aNode指向node
    # 然后node暂时为最后一个node需要指向None
    # def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
    #     if head == None:
    #         return None
    #     if head.next == None:
    #         return head
    #     newHead = self.reverseList(head.next)
    #     head.next.next = head
    #     head.next = None
    #     return newHead

    # 迭代，可以用双指针, prev,curr
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev, curr = None, head
        while curr != None:
            tmp = curr.next
            curr.next = prev
            prev = curr
            curr = tmp 
        return prev
    

    # 迭代 我的迭代方法
    # def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
    #     aNode = head
    #     dummy = ListNode()
    #     while aNode != None:
    #         suiv = aNode.next
    #         tmp = dummy.next
    #         dummy.next = aNode
    #         aNode.next = tmp
    #         aNode = suiv
    #     return dummy.next